azure/services/async/async.go

```diff
@@ -106,20 +106,23 @@ func (s *Service[C, D]) CreateOrUpdateResource(ctx context.Context, spec azure.R
 
 	result, poller, err := s.Creator.CreateOrUpdateAsync(ctx, spec, resumeToken, parameters)
 	errWrapped := errors.Wrapf(err, "failed to create or update resource %s/%s (service: %s)", rgName, resourceName, serviceName)
-	if poller != nil {
+	if poller != nil && azure.IsContextDeadlineExceededOrCanceledError(err) {
 		future, err := converters.PollerToFuture(poller, infrav1.PutFuture, serviceName, resourceName, rgName)
 		if err != nil {
 			return nil, errWrapped
 		}
 		s.Scope.SetLongRunningOperationState(future)
 		return nil, azure.WithTransientError(azure.NewOperationNotDoneError(future), requeueTime())
-	} else if err != nil {
-		return nil, errWrapped
 	}
 
-	// Once the operation is done, delete the long-running operation state.
+	// Once the operation is done, delete the long-running operation state. Even if the operation ended with
+	// an error, clear out any lingering state to try the operation again.
```

Comment:
this is the equivalent of what we were doing before https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/a521b7b0ce8c200519ed779e951db019130cde7d#diff-0e657fbf13cf152e97cf8871a3baf550199d64f99f94316e7e1b9eeb5d6cc8e4L90